e2e/require-after-teardown/__tests__/lateRequire.test.js

```diff
@@ -11,5 +11,5 @@ test('require after done', () => {
     const double = require('../');
 
     expect(double(5)).toBe(10);
-  }, 0);
+  }, 100);
```

Comment:
Are `0` still detected? If not, we should try to set some other flag so we can detect this case immediately after tests are done

Reply:
Yes, they are detected. That is why I had to update this test (and the other one as well).

Reply:
If it's still detected I'm not 100% sure why this had to change? Is the error just different?

Reply:
I took a deep loop into both cases (jest environment method access after teardown and late require) and both seems fragile already.

1. Any async event handler can break it. No matter what the even handler would do, anytime it takes at least one even-loop turnaround, this gets broken. More even-loop turnarounds in an async event handler, the more "suppressed" errors like this. Both errors are thrown on runtime level when the circus is out of game already. The runtime assumes the the time between a test and tearing runtime down is strictly sync (which may not be).
3. "jest environment method access after teardown" validates a single timers method access only. All other Jest environment methods usages are not reported at all, even on the current main branch.

I introduced the first truly async event handler. Therefore, the detection has changed because of the first point.

I was thinking about a new flag, however, I would need to make a hard choice as a user between the following variants:

A)
- I can keep the current behaviour detecting late require as well as late global method usage.
- I won't be able to detect unhandled rejections when the promise rejection happens in the same event loop turn as the test ends.
- I would still get false failures for unhandled rejections when the rejected promise handling happens in the same event loop turn as the test ends.
- I would still get global unhandled rejection error in case there is a floating promise being rejected way after its creation (e.g. when tearing down the env).

B)
- I get all promise rejections error accurately (except rejections happening way after the promise creation).
- I won't get reported on late requires nor late global method usage when it happens within the same event loop turn as the test ends (as it is still considered as part of the test).

I would prefer variant B) as you would probably guess. What is the reasoning behind those checks though?

I think there is a variant C) though. It would mean those 2 specific conditions get changed so they are detect on circus level rather than on runtime level. However, I am not sure how feasible that.

I was also thinking about variant D) - capturing all pending timers, leaving there the one waiting for the next even loop turn and the reapply the remaining timers. However, it feels hacky and I was not able to think of a way for other envs than Node.js (and even that is questionable).

Reply:
Thanks for digging! I think your change is fine considering the tradeoffs.

I guess what we could do was signalling from Circus that we are done running tests once the `teardown` event fires, at which point the runtime would start rejecting new `require`s etc. even though it hasn't started its own teardown. Might not really be feasible in the current architecture, tho?

Reply:
I actually got the same idea in the evening but I did not have time to look into it yet 😃 I will give it a shot

Reply:
I believe I have the solution. I updated test snapshots so whole output is in the snapshot as working with specific lines seems unnecessary and possibly problematic.